SFTP hook to prefer the SSH paramiko key over the key file path

30blay opened the pull request:
Related issue: [#16323](https://github.com/apache/airflow/issues/16323)

The SFTP hook does not allow for a SSH private key string to be used, only a private key file path. As it is based on the SSH hook, which does allow for a private key string to be used, a small change could align these two hooks.

This change first checks if a paramiko object is present after the SSH hook has been initialised, and the prefers that rather than the ssh key file path.

This PR replaces stale PR [#16338](https://github.com/apache/airflow/pull/16338)

hx-markterry committed:
Prefer the paramiko key over the key path

30blay committed:
Docs: add that private_key can the content of the private key

30blay committed:
Add SFTPHook test with private key content as string

30blay committed:
Revert "Prefer the paramiko key over the key path"

This reverts commit e610576e179c541c30dda8852ebdd063c8975679.

30blay committed:
Add tests

30blay committed:
Fix private_key argument

30blay committed:
Fix documentation

Remove the link to pysftp. SFTP Connection's private_key is the key content, while pysftp's private_key is a file path. Readers should look at the documentation for SFTP Connection for argument definitions

uranusjr commented on docs/apache-airflow-providers-sftp/connections/sftp.rst:
```diff
@@ -72,7 +70,8 @@ Extra (optional)
       the endpoint matches this value.
-    * ``private_key`` Specify the path to private key file(str) or paramiko.AgentKey
+    * ``private_key`` Specify the content of the private key, the path to the private key file(str) or paramiko.AgentKey
+    * ``key_file`` - Full Path of the private SSH Key file that will be used to connect to the remote_host.
```
Since `SFTPHook` really just takes all the extras that `SSHHook` does, maybe we should just tell the user to read the documentation page on SSH connections instead. Not quite sure.

30blay replied:
SFTPHook doesn't use some of the extras that SSHHook uses, so I think telling the user to read the SSH connection documentation would be misleading.
For example `timeout`

uranusjr replied:
Makes sense, thanks.

30blay committed:
Fix documentation examples

Since private_key is now the actual private key content, and key_file is not the path to the private key file

30blay committed:
Reformat SFTPHook tests

uranusjr reviewed (approved):
Assuming CI passes.
